### PR TITLE
Highlight groups for the HTML Stmt file and tooltips to reveal types.

### PR DESCRIPTION
**Pull request by mcourteaux**
While I was investigating the buffer index datatypes for #7884 myself, I added tooltips to reveal the datatype of the IR node in the Stmt file.

type tooltip

Another feature I wanted much, to analyze variables (especially when there is too much code to fit on the screen and you have to start scrolling, which makes the highlight-on-hover impractical) is to color matching variables by clicking them. There are 7 colors slots which will automatically recycle if you un-highlight a variable.

highlighting

Also minor bugfix in the printing: Reinterpret casts were printed as normal casts before.

### CODE REVIEW

**Comment by mcourteaux** [comment]
@antonysigma Feel free to run the linter over the new JS.

**Comment by antonysigma** [comment]
> @antonysigma Feel free to run the linter over the new JS.

For sure. Here you are:

```javascript
/home/antony/Projects/halide/src/irvisualizer/html_template_StmtToHTML.js
   29:9   error  'matchedElements' is already defined  no-redeclare
   44:21  error  'highlightIdx' is already defined     no-redeclare
   65:5   error  'makeCodeVisible' is not defined      no-undef
  100:39  error  'event' is defined but never used     no-unused-vars
  156:18  error  'i' is already defined                no-redeclare
  166:13  error  'resizer' is already defined          no-redeclare
  167:18  error  'i' is already defined                no-redeclare
  199:51  error  'device_code_Tab' is not defined      no-undef

✖ 8 problems (8 errors, 0 warnings)
```

Feel free to `git cherry-pick` this commit to reproduce the above `eslint` linter warnings.

https://github.com/antonysigma/halide/tree/javascript-scan-for-virus

**Comment by mcourteaux** [comment]
Cleaned up the code, all warnings are gone. This was actually useful! I retested the result, and all looks fine. Ready to merge.

File: src/StmtToHTML.cpp [hunk 1]
```diff
@@ -1070,11 +1080,17 @@ class HTMLCodePrinter : public IRVisitor {
         context_stack_tags.push_back(tag);
     }
 
-    void print_opening_tag(const std::string &tag, const std::string &cls, std::string id) {
-        stream << "<" << tag << " class='" << cls << "' id='" << id << "'>";
+    /*
```
Comment (abadams) [inline src/StmtToHTML.cpp:1083]: Please delete commented-out code, or add a comment that explains why it is preserved.

File: src/StmtToHTML.cpp [hunk 2]
```diff
@@ -1086,17 +1102,19 @@ class HTMLCodePrinter : public IRVisitor {
...
-        print_opening_tag(tag, cls, id);
+    void print_html_element(const std::string &tag, const std::string &cls, const std::string &body, const std::string &tooltip = "", int id = -1) {
+        print_opening_tag(tag, cls, tooltip, id);
         stream << body;
         print_closing_tag(tag);
     }
 
-    void print_html_element(const std::string &tag, const std::string &cls, const std::string &body, std::string id) {
-        print_opening_tag(tag, cls, id);
+    /*
```
Comment (abadams) [inline src/StmtToHTML.cpp:1111]: Same comment here.

File: src/StmtToHTML.cpp [hunk 3]
```diff
@@ -1542,10 +1574,15 @@ class HTMLCodePrinter : public IRVisitor {
     }
 
     void visit(const Call *op) override {
+        /*
```
Comment (abadams) [inline src/StmtToHTML.cpp:1577]: More commented-out code here.

**Comment by mcourteaux** [comment]
Oh sorry about those. I should make a habit of checking and reviewing the overall changes when I open a PR.

**Comment by mcourteaux** [comment]
Just to be clear, this is ready to merge. :slightly_smiling_face:

**Comment by abadams** [comment]
Could you do a merge with main? That will hopefully clear up most of the CI failures.

**Review by abadams — APPROVED** [review]

**Comment by mcourteaux** [comment]
Thanks for merging!  :partying_face:

I'm looking forward to hearing feedback on the HTML Stmt IR, as opposed to the plaintext Stmt IR files. I'm using them for a long time now, and I generally really like it. If there are obvious shortcomings, please let me know. I would love to see people start using this and prefer it. So if there's anything bothering/blocking you from using it, I'd be happy to hear it.

Here is [a demo file](https://silvernode.io/halide_stmt/neonraw_denoiser_f8_k5_p1.conceptual.stmt.html) I hosted for quickly checking it out.

Quick start: generator emit types are `stmt_html` and `conceptual_stmt_html`. Works great on `host`, and `host-cuda` targets. I get the best performance with Chromium, Firefox is significantly slower at loading the page.

 - @steven-johnson Maybe you can introduce this at Google if you think it's worth it?
 - @abadams @vksnk @maaz139 @antonysigma You were involved at some point in the discussion as well.

**Comment by vksnk** [comment]
This looks really cool and is indeed a big improvement over IR files!

> * @steven-johnson Maybe you can introduce this at Google if you think it's worth it?

I've pulled in the latest Halide changes into Google code base, so this should be available for internal use.  I think we will need to write a short how-to on how to use it with Google build environment and after that we can advertise it in local Halide user group (actually, just last week someone asked me if there is an easy way to see a generated assembly for inner loops and it seems to fit the bill perfectly).

**Comment by mcourteaux** [comment]
> I think we will need to write a short how-to on how to use it with Google build environment

Just as a note, it's not available through the Func or Pipeline C++ API, only through the Generators, which might not be a situation that applies to everyone.

**Comment by abadams** [comment]
Unless things have changed since I was there, Google is 100% Generators

**Comment by antonysigma** [comment]
@mcourteaux I am still catching up... Currently struggling with the emit option `-e conceptual_stmt_html` that results in `Unhandled exception: map::at`. I will workaround issue with `-e conceptual_stmt_html,html` and then checkout the new features in this PR.

**Comment by mcourteaux** [comment]
@antonysigma Sorry, this is the second time you report a bug like this. I fixed this one again. Hopefully all cases are covered. PR in #7914.

**Pull request merged**